### DISCUSSION TITLE
[Released] Git LFS

### DISCUSSION THREAD

**Discussion by drewwyatt** [vercel/community#49] [Idea]
is this already a thing? If it is, I just may not how to enable it, but all of the files I am storing in LFS are not being checked out in my deployments.

🙏 🙏 🙏 

_Update: Vercel has support for Git LFS now: https://vercel.com/changelog/git-lfs-support 🎉_

**Comment by johnchristopherjones** [comment]
Someone mentioned [proxying with serverless functions](https://spectrum.chat/zeit/now/git-lfs-large-file-storage~7e6b7a3a-53a5-41d8-bd25-32b36d288496), but I haven't yet worked the details of that out.

**Comment by goodhoko** [comment]
This is currently the only thing that keeps from migrating to Vercel. Is this feature is on the roadmap? If not, is there any workaround that would make the LFS files available during the build process? Since my app does an image processing during the build, proxying via serverless functions does not seem as a feasible solution.

**Comment by devklepacki** [comment]
There is already a one closed issue for this. For refference: https://github.com/vercel/vercel/issues/2783 Would love to see this feature tho

**Comment by srcnalt** [comment]
Just hit this issue right now, wondering when would this be available?

**Comment by masakudamatsu** [comment]
I hit this issue right now as well...

**Comment by riccardogiorato** [comment]
I hit this issue too, will need to switch to other hosting providers unluckily if there won't be a public workaround or fix to this.

**Comment by theCompanyDream** [comment]
I hit this issue too. Not using git lfs just really isn't an option with a monorepo I hope that this still get's attention in the future. Else I will also have to move away from Vercel.

**Comment by johnchristopherjones** [comment]
I've implemented the following workaround to make Git LFS files work with the Vercel CDN.

In short, I stick my Git LFS files in a particular directory and ask Vercel to redirect HTTP requests for that directory to my separate asset server.  According to Vercel's docs, this [redirect rule results in a HTTP 308 response](https://vercel.com/docs/configuration#project/redirects), which means it [triggers edge network caching](https://vercel.com/docs/edge-network/caching).

1. I deploy assets to a web server or CDN.  For example, to rsync to a vanilla static server, I merge the following into my package.json:
    ```json
    { "scripts": {
      "deploy:assets": "rsync -rvz --delete-after --force src/assets/ user@password:/srv/www/my_project/public_html/assets/"
    }
    ```
    I can then call that script to deploy assets.  You can do something very similar with most CDNs.  However, you really don't need a CDN since Vercel will cache after the first request.  I can just run this manually whenever, but I can also run it, or a variation of it, in a local git hook or remotely trigger it with a deployment hook via GitHub Actions etc.

2. I add a redirect to the server I deployed to in my `vercel.json`:
    ```json
    {
      "redirects": [
        { "source": "/assets/(.*)", "destination": "https://my_asset_server.my_project.com/assets/$1" }
      ]
    }
    ```

Now, if my site is at https://foo.com and I have an `<img src="/assets/logo.png"/>`, it gets redirected to my asset server.

There are two downsides to this approach:
1. Some additional setup.
2. In the described configuration, the assets are global, so they are shared among all preview branches and production.

The additional setup is pretty minimal.

The preview branch thing bothers me.  I think the setup could be refined by including a commit hash in the path in the redirect and deployment script from an environment variable.  I haven't needed this yet.  I haven't been able to determine from the docs if the vercel.json file supports environment variables in redirects.  Even if the redirect doesn't directly support it, you could definitely use a Vercel environment variable in your project to prefix to your asset paths.

**Comment by naavis** [comment]
I would be really happy to see Vercel support Git LFS with Github too. I'm building a pretty image-heavy site and the images are stored in Git LFS. Right now I am able to work around the limitation by building the site with Github Actions and just deploying a static site to Vercel. But it would of course be much more streamlined to be able to just throw the whole project at Vercel and let it do its thing.

**Reply by ghost** [reply]
Yeah I have the same issue. It actually took me a while to figure out that my image processing was failing because the images weren't being pulled in from git-lfs (there were zero errors mentioning that git-lfs doesn't work).

So yeah:

1. Clear error when cloning a repo with lfs references
2. Allowing git-lfs would be nice

**Comment by Southclaws** [comment]
I just spent ages migrating our site to use LFS because it was 600MB (on a slow connection too!) and found out it isn't supported 😂

Would be great to see this added!

**Comment by WilliamJohnDavisIII** [comment]
Just giving my +1 to this feature request. I spent 2 hours fiddling around until I realized the relationship (or lack thereof) between Vercel and Git LFS was the cause of my image links being broken.

I would love to at least see the lack of support mentioned in the docs, or as an error during build time.

**Comment by flogy** [comment]
Also just ran into this one ond would love to see Vercel supporting large files too.

**Comment by bravetheheat** [comment]
Migrating to Vercel, and we need this feature.

**Comment by naokazuterada** [comment]
+1 I want this feature seriously. Please!

**Comment by chrispoe** [comment]
+1. Yes, please!

**Comment by TaraOShea** [comment]
+ 1

**Comment by abstractalgo** [comment]
+1

**Comment by JotaPe** [comment]
+1

**Comment by WilliamJohnDavisIII** [comment]
-1

**Comment by WilliamJohnDavisIII** [comment]
jk I still want this

**Comment by unitof** [comment]
I'm playing around with using `routes` (and possibly later on a serverless function) to proxy requests to LFS'd assets back to GitHub's LFS CDN (which is `https://media.githubusercontent.com/media/[username]/[repo]/[branch]/[file/path.ext]`)

https://github.com/unitof/vercel-git-lfs-test/

[`routes`](https://github.com/unitof/vercel-git-lfs-test/blob/main/vercel.json#L2-L6) [works](https://vercel-git-lfs-test.vercel.app)!

The newer/better `rewrites` [_sorta_](https://github.com/unitof/vercel-git-lfs-test/commit/7cc47d69fcae3d7eebc89c33f48a8accbbe4a7e2) does, but if a file exists in the filesystem with the name, serving that file [takes priority over the `rewrites` rule](https://github.com/vercel/vercel/discussions/5303), which for LFS is always the case (and it ends up serving the [LFS pointer as an image](https://vercel-git-lfs-test-dyqvpq3f6.vercel.app/hat.jpg)).

Next up:
- [ ] see if Vercel caching headers work to cut down on GitHub requests (GitHub Data Packs include 50GB of bandwidth each, or 1GB free).
- [ ] See if I can write a serverless function to keep this properly in sync with `.gitattributes`, as well as pointing to the correct branch at GitHub's CDN for preview branches.
- [ ] Use case: eventually I want to build a Git LFS-based blogging engine capable of a full Git blame/history, so you can "checkout" an older version (including any embedded images) of any post at a commit hash-based URL.

**Comment by CatStudioApp** [comment]
Wonder whether there's still a plan? No git-lfs support sounds unbelievable in 2021.

**Comment by traveling-developer** [comment]
need this too

**Comment by GautierArcin** [comment]
bump

**Comment by amacneil** [comment]
I got it to work by overriding my install script to:

```
set -x && git remote add origin https://github.com/foo/bar.git && amazon-linux-extras install -y epel && yum install -y git-lfs && git lfs install && git lfs pull && yarn install
```

Caveat this is a public github repo, so I'm not sure how it would work if authentication is needed.

**Reply by brhx** [reply]
Wow that's perfect thank you so much! You're a life saver 🎉 

For those of you using this with a private repository, you can modify the install command (below) to use authentication with a [personal access token](https://github.com/settings/tokens):

```shell
set -x && git remote add origin https://{GITHUB_USERNAME}:{PERSONAL_ACCESS_TOKEN}@github.com/foo/bar && amazon-linux-extras install -y epel && yum install -y git-lfs && git lfs install && git lfs pull && yarn install
```

**Reply by looterz** [reply]
Thanks, this worked great as a temporary solution.

**Reply by ttoinou** [reply]
Anyone knows how to make this work with Gitlab ?

**Reply by Alwinator** [reply]
Unfortunately, I am getting the following error when `git lfs pull` is called. error: failed to fetch some objects from 'https://Alwinator:..@github.com/Alwinator/myproject.git/info/lfs' I am only getting this error in the Vercel build pipeline, locally it works without a problem. Any ideas?

**Reply by unitof** [reply]
@Alwinator Just to confirm: your repo (and LFS) is public? If so would you mind sharing the real URL?

**Reply by Alwinator** [reply]
@unitof Unfortunately, my repo is private. However, I tried exactly the same commands on my local computer and they worked.

**Reply by DwarfThema** [reply]
What does it mean to override the install script?
Which directory should I change the script contents of which file?
I am using next.js.

**Reply by amyegan** [reply]
You can override the default Install Command in your Project Settings under Build & Development Settings

<img width="944" alt="Image of changed install command" src="https://user-images.githubusercontent.com/8247105/191030790-3bc7996e-45e0-470a-9495-7c56c18bf0e0.png">

**Comment by ghost** [comment]
Especially now with next/image, which allow for a lot of optimizations if images are local to the app, it would be very nice to have git lfs support baked into the image used to build our next apps on vercel.

**Comment by eAi** [comment]
We could really do with this too - think we may need to turn off git integration without it.

**Comment by jackbach** [comment]
Need this feature too!

**Comment by simonwidjaja** [comment]
Need this feature too!

**Comment by snowzurfer** [comment]
Need it too!

**Comment by ttoinou** [comment]
I need this too ! How do people even build websites with Vercel without big static files storage ? What are our options here ?

**Reply by matootie** [reply]
For now use this recommendation https://github.com/vercel/vercel/discussions/3716#discussioncomment-1395097

Just remember to change `{GITHUB_USERNAME}` to the account username, change `{PERSONAL_ACCESS_TOKEN}` to a token that has read permissions on repos, and change the repo url "`github.com/foo/bar`" to the repo url for your project. It works for me on a number of sites I run on Vercel.

I still want an integrated solution though >:(

**Reply by ttoinou** [reply]
Thanks but we tried and it doesn't work. We use gitlab, not github. Also, we don't have the error output from Vercel so we don't know how to fix it

**Comment by Stanzilla** [comment]
I really like how Netlify handles it https://docs.netlify.com/large-media/setup/

**Reply by jackbach** [reply]
Note that Netlify is a bit confusing with this.
Netlify Large Media is a service to store big files in Netlify (which uses LFS)
Alternatively, it is possible to use Github for LFS by setting a couple of not really obvious env variables (GIT_LFS_ENABLED and GIT_LFS_FETCH_INCLUDE) documented here](https://docs.netlify.com/configure-builds/environment-variables/#netlify-configuration-variables).
I would rather have a solution like the later. Or even better, make LFS work out of the box by checking for .gitattributes or other lfs related config files in the repo.

**Reply by Stanzilla** [reply]
> Note that Netlify is a bit confusing with this. Netlify Large Media is a service to store big files in Netlify (which uses LFS) Alternatively, it is possible to use Github for LFS by setting a couple of not really obvious env variables (GIT_LFS_ENABLED and GIT_LFS_FETCH_INCLUDE) documented here]([docs.netlify.com/configure-builds/environment-variables/#netlify-configuration-variables](https://docs.netlify.com/configure-builds/environment-variables/#netlify-configuration-variables)). I would rather have a solution like the later. Or even better, make LFS work out of the box by checking for .gitattributes or other lfs related config files in the repo.

Yes that is a very important point! I wasn't aware that you could LFS with them without using their service.

**Reply by janek** [reply]
Thanks for writing this, saved me some time. I submitted feedback to Netlify docs suggesting to make that more explicit there.

**Comment by btakita** [comment]
Is this your most requested feature? It sure would be great to have git lfs support

**Comment by amyegan** [comment]
We don't have automatic support for Git LFS, but it can still be used with a configuration change. I want to highlight the contributions of other community members and note that you can use Git LFS by including it in your Install Command. https://github.com/vercel/community/discussions/49#discussioncomment-739107

I'll make sure the request for integrated Git LFS support is put in front of the right people to investigate

**Reply by mirshko** [reply]
Regarding that config change, is it possible to encrypt the PAT needed to add the new origin for this? 

Using an env var works but the result from the `git add origin` prints the PAT after it finishes running. 

Of course native support would be awesome!

**Reply by tribixbite** [reply]
> We don't have automatic support for Git LFS, but it can still be used with a configuration change. I want to highlight the contributions of other community members and note that you can use Git LFS by including it in your Install Command. [#49 (comment)](https://github.com/orgs/vercel/discussions/49#discussioncomment-739107)
> 
> I'll make sure the request for integrated Git LFS support is put in front of the right people to investigate

It would be great if you could increase the character limit for the custom install command field. It's currently set to 256. With the new Fine-grained personal access tokens the above command ends up being about 298 characters.

**Reply by matootie** [reply]
> > We don't have automatic support for Git LFS, but it can still be used with a configuration change. I want to highlight the contributions of other community members and note that you can use Git LFS by including it in your Install Command. [#49 (comment)](https://github.com/orgs/vercel/discussions/49#discussioncomment-739107)
> 
> > 
> 
> > I'll make sure the request for integrated Git LFS support is put in front of the right people to investigate
> 
> 
> 
> It would be great if you could increase the character limit for the custom install command field. It's currently set to 256. With the new Fine-grained personal access tokens the above command ends up being about 298 characters.

What I switched to recently is a setup where you build your app within GitHub Actions (very easy to pull from LFS there), and then deploy the app to Vercel using the CLI and an access token from your Vercel account.

~~I don't have access to the exact link now because I'm on GitHub mobile but if you do a search for "Vercel GitHub Actions" you should find a guide from their website.~~
Update: it should be this link [https://vercel.com/guides/how-can-i-use-github-actions-with-vercel](https://vercel.com/guides/how-can-i-use-github-actions-with-vercel)

It takes some extra time to configure but I think it's the cleanest alternative to the above mentioned install-script hack.

**Comment by Spiralis** [comment]
So, LFS is still not supported? I hear about the workarounds. But, seriously, they are "workarounds". A warning or error during build would at least give a hint about the problem...

**Comment by tendiesman** [comment]
Hey this override
`set -x && git remote add origin https://github.com/tendiesman/gm_cafe.git && amazon-linux-extras install -y epel && yum install -y git-lfs && git lfs install && git lfs pull && yarn install`
Didn't work for me. Public repo and domain is gm.cafe ; could someone check on what I might be doing wrong?

**Reply by tendiesman** [reply]
Nvm, for future reference, make sure you have pushed any updates to .gitattributes

**Comment by pablorodrigo** [comment]
Something new about it? I have tried to use some of the workarounds above, but with no success.

**Comment by parisholley** [comment]
unfortunately the way the vercel build pipeline is setup, not sure this is possible. there are a few issues you gotta fix before you can use git lfs:

- the local checkout does not have a remote, so it needs to be added
- github (or whoever your git provider is) will need to have its  ssh host verification key added since it isn't in the build image
- then you can install the git-lfs software as noted above

in theory those points above can be solved using this install command:

```ssh-keyscan -t rsa github.com >> ~/.ssh/known_hosts && git remote add origin git@github.com:MY_ORG/MY_REPO.git && amazon-linux-extras install -y epel -y && yum install git-lfs && git lfs install && git lfs pull && yarn install``` 

even if you get that command to run, you will likely hit an authentication problem since vercel isn't including a private key in the build image (rightfully so to protect exposing access to repo from build script). if you can find a secure way to generate an ssh key and add it to your repo as well as inject it into the build process in a secure way, then it might work.

i have been using git-lfs to store my `.yarn/cache` folder, which won't work using vercel ci. only options are to either:

- don't use git lfs :)
- in the case of `.yarn/cache`, `rm -rf .yarn/cache && yarn install` to clean the empty files and force a download
- use a different cache folder location (sames con as `rm -rf` in that you loose benefit of zero install if NPM goes crazy, but has benefit of vercel caching between builds) `YARN_CACHE_FOLDER=./.next/cache/yarn-cache yarn install` (see discussion here: https://github.com/orgs/vercel/discussions/222)
- build using an external ci tool (eg: github actions) and only use vercel to deploy
- vercel would need to update their pipeline to do a git lfs pull on their side before copying files into the build image

**note:** the commands above will need to be adjusted if you are not in the "Root Directory", eg:
- `rm -rf ../../.yarn/cache`

**Comment by leerob** [comment]
We're working on support for this. Stay tuned 🙏

**Comment by amyegan** [comment]
We now have [Git LFS](https://vercel.com/docs/concepts/projects/overview#git-large-file-storage-lfs) support! 🙌 https://vercel.com/changelog/git-lfs-support 🎉

**Reply by eAi** [reply]
Amazing!
